File: go/src/enjarify-go/jvm/cpool/constantpool.go
// Copyright 2015 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package cpool

import (
	"enjarify-go/byteio"
	"enjarify-go/dex"
	"enjarify-go/jvm/errors"
	"enjarify-go/util"
)

const CONSTANT_Class = 7
const CONSTANT_Fieldref = 9
const CONSTANT_Methodref = 10
const CONSTANT_InterfaceMethodref = 11
const CONSTANT_String = 8
const CONSTANT_Integer = 3
const CONSTANT_Float = 4
const CONSTANT_Long = 5
const CONSTANT_Double = 6
const CONSTANT_NameAndType = 12
const CONSTANT_Utf8 = 1

// const CONSTANT_MethodHandle = 15
// const CONSTANT_MethodType = 16
// const CONSTANT_InvokeDynamic = 18
const MAX_CONST = CONSTANT_NameAndType

func width(tag byte) int {
	if tag == CONSTANT_Double || tag == CONSTANT_Long {
		return 2
	}
	return 1
}

type Data struct {
	s      string
	p1, p2 uint16
	X      uint64
}

type Pair struct {
	Tag byte
	Data
}

type subclassImpl interface {
	Vals() []Pair
	getInd(low bool, width int) uint16
	space() int
	lowspace() int
}

type constantPoolBase struct {
	sub    subclassImpl
	lookup [MAX_CONST + 1]map[Data]uint16
}

func newConstantPoolBase() constantPoolBase {
	self := constantPoolBase{}
	for i := 0; i < len(self.lookup); i++ {
		self.lookup[i] = make(map[Data]uint16)
	}
	return self
}

func (self *constantPoolBase) get(tag byte, data Data) uint16 {
	d := self.lookup[tag]
	if val, ok := d[data]; ok {
		return val
	}

	low := (tag == CONSTANT_Integer || tag == CONSTANT_Float || tag == CONSTANT_String)
	index := self.sub.getInd(low, width(tag))
	d[data] = index
	self.sub.Vals()[index] = Pair{tag, data}
	return index
}

func (self *constantPoolBase) InsertDirectly(pair Pair, low bool) {
	d := self.lookup[pair.Tag]
	index := self.sub.getInd(low, width(pair.Tag))
	d[pair.Data] = index
	self.sub.Vals()[index] = pair
}

func (self *constantPoolBase) TryGet(pair Pair) (index uint16, ok bool) {
	tag := pair.Tag
	data := pair.Data

	d := self.lookup[tag]
	if val, ok := d[data]; ok {
		return val, true
	}

	width := width(tag)
	if width > self.sub.space() {
		return 0, false
	}

	index = self.sub.getInd(true, width)
	d[data] = index
	self.sub.Vals()[index] = pair
	return index, true
}

func (self *constantPoolBase) Space() int {
	return self.sub.space()
}

func (self *constantPoolBase) LowSpace() int {
	return self.sub.lowspace()
}
func (self *constantPoolBase) Utf8(s string) uint16 {
	if len(s) > 65535 {
		panic(&errors.ClassfileLimitExceeded{})
	}
	return self.get(CONSTANT_Utf8, Data{s: s})
}

func (self *constantPoolBase) Class(s string) uint16 {
	return self.get(CONSTANT_Class, Data{p1: self.Utf8(s)})
}

func (self *constantPoolBase) String(s string) uint16 {
	return self.get(CONSTANT_String, Data{p1: self.Utf8(s)})
}

func (self *constantPoolBase) Nat(name, desc string) uint16 {
	return self.get(CONSTANT_NameAndType, Data{p1: self.Utf8(name), p2: self.Utf8(desc)})
}

func (self *constantPoolBase) triple(tag byte, trip dex.Triple) uint16 {
	return self.get(tag, Data{p1: self.Class(trip.Cname), p2: self.Nat(trip.Name, trip.Desc)})
}

func (self *constantPoolBase) Field(trip dex.Triple) uint16 {
	return self.triple(CONSTANT_Fieldref, trip)
}

func (self *constantPoolBase) Method(trip dex.Triple) uint16 {
	return self.triple(CONSTANT_Methodref, trip)
}

func (self *constantPoolBase) IMethod(trip dex.Triple) uint16 {
	return self.triple(CONSTANT_InterfaceMethodref, trip)
}

func (self *constantPoolBase) Int(x uint32) uint16 {
	return self.get(CONSTANT_Integer, Data{X: uint64(x)})
}

func (self *constantPoolBase) Float(x uint32) uint16 {
	return self.get(CONSTANT_Float, Data{X: uint64(x)})
}

func (self *constantPoolBase) Long(x uint64) uint16 {
	return self.get(CONSTANT_Long, Data{X: x})
}

func (self *constantPoolBase) Double(x uint64) uint16 {
	return self.get(CONSTANT_Double, Data{X: x})
}

func (self *constantPoolBase) writeEntry(stream *byteio.Writer, item Pair) {
	if item.Tag == 0 {
		return
	}

	stream.U8(item.Tag)
	switch item.Tag {
	case CONSTANT_Utf8:
		stream.U16(uint16(len(item.s)))
		stream.WriteString(item.s)
	case CONSTANT_Integer, CONSTANT_Float:
		stream.U32(uint32(item.X))
	case CONSTANT_Long, CONSTANT_Double:
		stream.U64(item.X)
	case CONSTANT_Class, CONSTANT_String:
		stream.U16(item.p1)
	default:
		stream.U16(item.p1)
		stream.U16(item.p2)
	}
}

type simpleConstantPool struct {
	constantPoolBase
	_vals []Pair
}

func (self *simpleConstantPool) Vals() []Pair {
	return self._vals
}

func (self *simpleConstantPool) space() int {
	return 65535 - len(self._vals)
}

func (self *simpleConstantPool) lowspace() int {
	return 256 - len(self._vals)
}

func (self *simpleConstantPool) getInd(low bool, width int) (index uint16) {
	if self.space() < width {
		panic(&errors.ClassfileLimitExceeded{})
	}

	index = uint16(len(self._vals))
	for i := 0; i < width; i++ {
		self._vals = append(self._vals, Pair{})
	}
	return
}

func (self *simpleConstantPool) Write(stream *byteio.Writer) {
	util.Assert(len(self._vals) <= 65535)
	stream.U16(uint16(len(self._vals)))
	for _, item := range self._vals {
		self.writeEntry(stream, item)
	}
}

var PLACEHOLDER_ENTRY = byteio.BH(CONSTANT_Utf8, 0)

type splitConstantPool struct {
	constantPoolBase
	bot, top int
	_vals    [65535]Pair
}

func (self *splitConstantPool) Vals() []Pair {
	return self._vals[:]
}

func (self *splitConstantPool) space() int {
	return self.top - self.bot
}

func (self *splitConstantPool) lowspace() int {
	return 256 - self.bot
}

func (self *splitConstantPool) getInd(low bool, width int) (index uint16) {
	if self.space() < width {
		panic(&errors.ClassfileLimitExceeded{})
	}

	if low {
		self.bot += width
		return uint16(self.bot - width)
	}
	self.top -= width
	return uint16(self.top)
}

func (self *splitConstantPool) Write(stream *byteio.Writer) {
	stream.U16(uint16(len(self._vals)))

	for _, item := range self._vals[:self.bot] {
		self.writeEntry(stream, item)
	}

	for _, _ = range self._vals[self.bot:self.top] {
		if _, err := stream.WriteString(PLACEHOLDER_ENTRY); err != nil {
			panic(err)
		}
	}

	for _, item := range self._vals[self.top:] {
		self.writeEntry(stream, item)
	}
}

type Pool interface {
	Vals() []Pair
	InsertDirectly(pair Pair, low bool)
	TryGet(pair Pair) (index uint16, ok bool)
	Space() int
	LowSpace() int

	Utf8(s string) uint16
	Class(s string) uint16
	String(s string) uint16
	Nat(name, desc string) uint16
	Field(trip dex.Triple) uint16
	Method(trip dex.Triple) uint16
	IMethod(trip dex.Triple) uint16
	Int(x uint32) uint16
	Float(x uint32) uint16
	Long(x uint64) uint16
	Double(x uint64) uint16

	Write(stream *byteio.Writer)
}

func Simple() Pool {
	pool := simpleConstantPool{newConstantPoolBase(), make([]Pair, 1)}
	pool.sub = &pool
	return &pool
}

func Split() Pool {
	pool := splitConstantPool{constantPoolBase: newConstantPoolBase(), bot: 1, top: 65535}
	pool.sub = &pool
	return &pool
}

File: enjarify-rs/src/jvm/constantpool.rs
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
use std::collections::HashMap;

use strings::*;
use byteio::Writer;
use dex::{FieldId, MethodId};
use error;

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
pub struct ArgsUtf<'a>(&'a bstr);
#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
pub struct ArgsInd(u16);
#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
pub struct ArgsInd2(u16, u16);
#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
pub struct ArgsPrim(pub u64);

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
pub enum Entry<'a> {
    Class(ArgsInd),
    Fieldref(ArgsInd2),
    Methodref(ArgsInd2),
    InterfaceMethodref(ArgsInd2),
    JString(ArgsInd),
    Integer(ArgsPrim),
    Float(ArgsPrim),
    Long(ArgsPrim),
    Double(ArgsPrim),
    NameAndType(ArgsInd2),
    Utf8(ArgsUtf<'a>),
}
pub use self::Entry::*;
impl<'a> Entry<'a> {
    fn width(&self) -> usize {
        match self {
            &Long(_) | &Double(_) => 2,
            _ => 1,
        }
    }

    // to sort matching the python order
    pub fn cmp_key(&self) -> (u8, u64) {
        match self {
            &Integer(ref data) => (3, data.0),
            &Float(ref data) => (4, data.0),
            &Long(ref data) => (5, data.0),
            &Double(ref data) => (6, data.0),
            _ => unreachable!(),
        }
    }
}

pub trait ConstantPool<'a> {
    // abstract methods
    fn space(&self) -> usize;
    fn lowspace(&self) -> usize;
    fn write(&self, stream: &mut Writer);
    fn get_ind(&mut self, low: bool, width: usize) -> u16;
    fn lookup(&mut self) -> &mut HashMap<Entry<'a>, u16>;
    fn vals(&mut self) -> &mut [Option<Entry<'a>>];

    // derived methods
    fn get(&mut self, entry: Entry<'a>) -> u16 {
        // println!("{} {} cp.get {:?}", self.lookup().len(), self.vals().len(), &entry);
        if let Some(val) = self.lookup().get(&entry) {
            return *val;
        }

        let low = match &entry {
            &Integer(_) | &Float(_) | &JString(_) => true,
            _ => false,
        };
        let index = self.get_ind(low, entry.width());
        self.lookup().insert(entry.clone(), index);
        self.vals()[index as usize] = Some(entry);
        index
    }

    fn insert_directly(&mut self, entry: Entry<'a>, low: bool) -> u16 {
        let index = self.get_ind(low, entry.width());
        self.lookup().insert(entry.clone(), index);
        self.vals()[index as usize] = Some(entry);
        index
    }

    fn try_get(&mut self, entry: Entry<'a>) -> Option<u16> {
        if let Some(val) = self.lookup().get(&entry) {
            return Some(*val);
        }

        if entry.width() > self.space() { return None; }
        Some(self.insert_directly(entry, true))
    }

    fn utf8(&mut self, s: &'a bstr) -> u16 {
        if s.len() > 65535 {
            error::classfile_limit_exceeded();
        }
        self.get(Utf8(ArgsUtf(s)))
    }

    fn class(&mut self, s: &'a bstr) -> u16 {
        let ind = self.utf8(s);
        self.get(Class(ArgsInd(ind)))
    }

    fn string(&mut self, s: &'a bstr) -> u16 {
        let ind = self.utf8(s);
        self.get(JString(ArgsInd(ind)))
    }

    fn _nat(&mut self, name: &'a bstr, desc: &'a bstr) -> u16 {
        let ind = self.utf8(name);
        let ind2 = self.utf8(desc);
        self.get(NameAndType(ArgsInd2(ind, ind2)))
    }

    fn field(&mut self, trip: &FieldId<'a>) -> u16 {
        let ind = self.class(trip.cname);
        let ind2 = self._nat(trip.name, trip.desc);
        self.get(Fieldref(ArgsInd2(ind, ind2)))
    }

    fn method(&mut self, trip: MethodId<'a>) -> u16 {
        let ind = self.class(trip.cname);
        let ind2 = self._nat(trip.name, trip.desc);
        self.get(Methodref(ArgsInd2(ind, ind2)))
    }

    fn imethod(&mut self, trip: MethodId<'a>) -> u16 {
        let ind = self.class(trip.cname);
        let ind2 = self._nat(trip.name, trip.desc);
        self.get(InterfaceMethodref(ArgsInd2(ind, ind2)))
    }

    fn int(&mut self, x: u32) -> u16 { self.get(Integer(ArgsPrim(x as u64))) }
    fn float(&mut self, x: u32) -> u16 { self.get(Float(ArgsPrim(x as u64))) }
    fn long(&mut self, x: u64) -> u16 { self.get(Long(ArgsPrim(x))) }
    fn double(&mut self, x: u64) -> u16 { self.get(Double(ArgsPrim(x))) }

    fn write_entry(&self, stream: &mut Writer, entry: &Option<Entry<'a>>) {
        if let Some(entry) = entry.as_ref() {
            match entry {
                &Class(ref args) => { stream.u8(7); stream.u16(args.0); },
                &Fieldref(ref args) => { stream.u8(9); stream.u16(args.0); stream.u16(args.1); },
                &Methodref(ref args) => { stream.u8(10); stream.u16(args.0); stream.u16(args.1); },
                &InterfaceMethodref(ref args) => { stream.u8(11); stream.u16(args.0); stream.u16(args.1); },
                &JString(ref args) => { stream.u8(8); stream.u16(args.0); },
                &Integer(ref args) => { stream.u8(3); stream.u32(args.0 as u32); },
                &Float(ref args) => { stream.u8(4); stream.u32(args.0 as u32); },
                &Long(ref args) => { stream.u8(5); stream.u64(args.0); },
                &Double(ref args) => { stream.u8(6); stream.u64(args.0); },
                &NameAndType(ref args) => { stream.u8(12); stream.u16(args.0); stream.u16(args.1); },
                &Utf8(ref args) => { stream.u8(1);
                    stream.u16(args.0.len() as u16);
                    stream.write(args.0);
                },
            }
        }
    }
}

// A simple constant pool that just allocates slots in increasing order.
struct SimplePool<'a> {
    lookup: HashMap<Entry<'a>, u16>,
    vals: Vec<Option<Entry<'a>>>,
}
impl<'a> ConstantPool<'a> for SimplePool<'a> {
    fn lookup(&mut self) -> &mut HashMap<Entry<'a>, u16> { &mut self.lookup }
    fn vals(&mut self) -> &mut [Option<Entry<'a>>] { &mut self.vals }

    fn space(&self) -> usize { 65535 - self.vals.len() }
    fn lowspace(&self) -> usize { 256usize.saturating_sub(self.vals.len()) }

    fn get_ind(&mut self, _low: bool, width: usize) -> u16 {
        if self.space() < width { error::classfile_limit_exceeded(); }
        let temp = self.vals.len();
        for _ in 0..width { self.vals.push(None); }
        temp as u16
    }
    fn write(&self, stream: &mut Writer) {
        stream.u16(self.vals.len() as u16);
        for item in &self.vals { self.write_entry(stream, item); }
    }
}

pub fn simple_pool<'a>() -> Box<ConstantPool<'a> + 'a> {
    Box::new(SimplePool{
        lookup: HashMap::new(),
        vals: vec![None],
    })
}

// Constant pool slots 1-255 are special because they can be referred to by the
// two byte ldc instruction (as opposed to 3 byte ldc_w/ldc2_w). Therefore, it is
// desireable to allocate constants which could use ldc in the first 255 slots,
// while not wasting these valuable low slots with pool entries that can't use
// ldc (utf8s, longs, etc.)
// One possible approach is to allocate the ldc entries starting at 1 and the
// others starting at 256, (possibly leaving a gap if there are less than 255 of
// the former). However, this is not ideal because the empty slots are not
// continguous. This means that you could end up in the sitatuation where there
// are exactly two free slots and you wish to add a long/double entry but the
// free slots are not continguous.
// To solve this, we take a different approach - always create the pool as the
// largest possible size (65534 entries) and allocate the non-ldc constants
// starting from the highest index and counting down. This ensures that the free
// slots are always contiguous. Since the classfile representation doesn't
// actually allow gaps like that, the empty spaces if any are filled in with
// dummy entries at the end.
// For simplicity, we always allocate ints, floats, and strings in the low entries
// and everything else in the high entries, regardless of whether they are actaully
// referenced by a ldc or not. (see ConstantPoolBase._get)

// Fill in unused space with shortest possible item (Utf8 ''), preencoded for efficiency
const PLACEHOLDER_ENTRY: &'static bstr = b"\x01\0\0";
struct SplitPool<'a> {
    lookup: HashMap<Entry<'a>, u16>,
    // vals: [Option<Entry<'a>>; 65535],
    vals: Vec<Option<Entry<'a>>>,
    bot: usize,
    top: usize,
}
impl<'a> ConstantPool<'a> for SplitPool<'a> {
    fn lookup(&mut self) -> &mut HashMap<Entry<'a>, u16> { &mut self.lookup }
    fn vals(&mut self) -> &mut [Option<Entry<'a>>] { &mut self.vals }

    fn space(&self) -> usize { self.top - self.bot }
    fn lowspace(&self) -> usize { 256usize.saturating_sub(self.bot) }

    fn get_ind(&mut self, low: bool, width: usize) -> u16 {
        if self.space() < width { error::classfile_limit_exceeded(); }
        (if low {
            self.bot += width;
            self.bot - width
        } else {
            self.top -= width;
            self.top
        }) as u16
    }
    fn write(&self, stream: &mut Writer) {
        stream.u16(65535);
        assert!(self.bot <= self.top);

        for item in &self.vals[..self.bot] { self.write_entry(stream, item); }

        stream.0.reserve(PLACEHOLDER_ENTRY.len() * (self.top - self.bot));
        for _ in 0..(self.top - self.bot) { stream.0.extend_from_slice(PLACEHOLDER_ENTRY); }

        for item in &self.vals[self.top..] { self.write_entry(stream, item); }
    }
}

pub fn split_pool<'a>() -> Box<ConstantPool<'a> + 'a> {
    Box::new(SplitPool{
        lookup: HashMap::new(),
        vals: vec![None; 65535],
        // vals: [None; 65535],
        bot: 1,
        top: 65535,
    })
}
